File: landrop-main/receiver/android/app/src/main/java/com/landrop/receiver/protocol/MessageStream.kt
package com.landrop.receiver.protocol

import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction

// Message boundary: transport bytes -> frames -> codec -> protocol messages.
// This is where the abstract Transport meets the Phase 02 codec. It is
// transport-agnostic: callers feed bytes (e.g. from Transport.read) and read
// complete ControlMessages out the other side.

sealed class JsonMessageResult {
    data class Message(val message: ControlMessage) : JsonMessageResult()
    object Incomplete : JsonMessageResult()
    data class FrameError(val problem: FrameProblem) : JsonMessageResult()
    data class ProtocolError(val problem: ProtocolProblem) : JsonMessageResult()
}

sealed class JsonFrameEncodeResult {
    data class Ok(val bytes: ByteArray) : JsonFrameEncodeResult()
    data class Failing(val problem: ProtocolProblem) : JsonFrameEncodeResult()
}

class JsonMessageStream(
    private val limits: Limits = Limits(),
    maxFramePayloadBytes: Int = MAX_FRAME_PAYLOAD_BYTES,
) {
    private val decoder = FrameDecoder(maxFramePayloadBytes)

    fun feed(chunk: ByteArray) = decoder.feed(chunk)

    fun readMessage(): JsonMessageResult = when (val frame = decoder.tryReadFrame()) {
        is FrameResult.Incomplete -> JsonMessageResult.Incomplete
        is FrameResult.Error -> JsonMessageResult.FrameError(frame.problem)
        is FrameResult.Frame -> {
            // Non-UTF-8 payload: normalize into the codec's malformed-JSON
            // problem rather than leaking a raw decoder exception.
            val text = strictUtf8(frame.payload)
                ?: return JsonMessageResult.ProtocolError(ProtocolProblem(ErrorCode.INVALID_REQUEST, "json"))
            when (val decoded = ProtocolJson.decode(text, limits)) {
                is Result.Success -> JsonMessageResult.Message(decoded.value)
                is Result.Failure -> {
                    val cause = decoded.exceptionOrNull()
                    if (cause is ProtocolProblem) {
                        JsonMessageResult.ProtocolError(cause)
                    } else {
                        JsonMessageResult.ProtocolError(ProtocolProblem(ErrorCode.INVALID_REQUEST, "json"))
                    }
                }
            }
        }
    }

    fun endOfStream(): StreamEndResult = decoder.endOfStream()

    val bufferedBytes: Long
        get() = decoder.bufferedBytes

    val isBalanced: Boolean
        get() = decoder.isBalanced
}

// Strict UTF-8 with reporting: malformed byte sequences yield null instead of
// replacement characters, matching the TS and Rust decoders.
private fun strictUtf8(bytes: ByteArray): String? = try {
    Charsets.UTF_8.newDecoder()
        .onMalformedInput(CodingErrorAction.REPORT)
        .onUnmappableCharacter(CodingErrorAction.REPORT)
        .decode(ByteBuffer.wrap(bytes))
        .toString()
} catch (_: java.nio.charset.CharacterCodingException) {
    null
}

// Encode a typed message into a single wire frame (JSON payload). The codec
// already bounds the JSON at maxJsonBytes, which equals the frame payload
// limit, so a validated message always fits one frame.
fun encodeJsonFrame(message: ControlMessage, limits: Limits = Limits()): JsonFrameEncodeResult =
    when (val json = ProtocolJson.encode(message, limits)) {
        is Result.Success -> {
            val encoded = encodeFrame(json.value.toByteArray(Charsets.UTF_8))
            when (encoded) {
                is FrameEncodeResult.Ok -> JsonFrameEncodeResult.Ok(encoded.bytes)
                is FrameEncodeResult.Failing -> {
                    // Unreachable in practice: validated JSON is bounded by maxJsonBytes.
                    JsonFrameEncodeResult.Failing(ProtocolProblem(ErrorCode.INVALID_REQUEST, "json.size"))
                }
            }
        }
        is Result.Failure -> JsonFrameEncodeResult.Failing(
            json.exceptionOrNull() as? ProtocolProblem
                ?: ProtocolProblem(ErrorCode.INVALID_REQUEST, "json"),
        )
    }

File: landrop-main/sender/desktop/src/protocol/message-stream.ts
// Message boundary: transport bytes -> frames -> codec -> protocol messages.
// This is where the abstract Transport meets the Phase 02 codec. It is
// transport-agnostic: callers feed bytes (e.g. from Transport.read) and read
// complete ControlMessages out the other side.

import { decodeMessage, encodeMessage, ProtocolProblem, DEFAULT_LIMITS } from "./codec";
import type { Limits } from "./codec";
import { FrameDecoder, encodeFrame, MAX_FRAME_PAYLOAD_BYTES } from "./framing";
import type { FrameProblem } from "./framing";
import type { ControlMessage } from "./types";

export type JsonMessageResult =
  | { readonly outcome: "message"; readonly message: ControlMessage }
  | { readonly outcome: "incomplete" }
  | { readonly outcome: "frame-error"; readonly problem: FrameProblem }
  | { readonly outcome: "protocol-error"; readonly problem: ProtocolProblem }
  | { readonly outcome: "end" };

export type StreamEndResult =
  | { readonly outcome: "ok" }
  | { readonly outcome: "frame-error"; readonly problem: FrameProblem };

export class JsonMessageStream {
  private readonly decoder: FrameDecoder;

  constructor(
    private readonly limits: Limits = DEFAULT_LIMITS,
    maxFramePayloadBytes: number = MAX_FRAME_PAYLOAD_BYTES,
  ) {
    this.decoder = new FrameDecoder(maxFramePayloadBytes);
  }

  feed(chunk: Uint8Array): void {
    this.decoder.feed(chunk);
  }

  readMessage(): JsonMessageResult {
    const frame = this.decoder.tryReadFrame();
    switch (frame.outcome) {
      case "incomplete":
        return { outcome: "incomplete" };
      case "error":
        return { outcome: "frame-error", problem: frame.problem };
      default: {
        let text: string;
        try {
          text = new TextDecoder("utf-8", { fatal: true }).decode(frame.payload);
        } catch {
          // Non-UTF-8 payload: normalize into the codec's malformed-JSON
          // problem rather than leaking a raw decoder exception (§10).
          return { outcome: "protocol-error", problem: new ProtocolProblem("INVALID_REQUEST", "json") };
        }
        try {
          return { outcome: "message", message: decodeMessage(text, this.limits) };
        } catch (cause) {
          if (cause instanceof ProtocolProblem) {
            return { outcome: "protocol-error", problem: cause };
          }
          throw cause;
        }
      }
    }
  }

  endOfStream(): StreamEndResult {
    const result = this.decoder.endOfStream();
    if (result.outcome === "error") {
      return { outcome: "frame-error", problem: result.problem };
    }
    return result;
  }

  get bufferedBytes(): number {
    return this.decoder.bufferedBytes;
  }

  get isBalanced(): boolean {
    return this.decoder.isBalanced;
  }
}

export type JsonFrameEncodeResult =
  | { readonly ok: true; readonly bytes: Uint8Array }
  | { readonly ok: false; readonly problem: ProtocolProblem | FrameProblem };

// Encode a typed message into a single wire frame (JSON payload). The codec
// already bounds the JSON at maxJsonBytes, which equals the frame payload
// limit, so a validated message always fits one frame.
export function encodeJsonFrame(
  message: ControlMessage,
  limits: Limits = DEFAULT_LIMITS,
): JsonFrameEncodeResult {
  try {
    const json = encodeMessage(message, limits);
    const encoded = encodeFrame(new TextEncoder().encode(json));
    if (!encoded.ok) return { ok: false, problem: encoded.problem };
    return { ok: true, bytes: encoded.bytes };
  } catch (cause) {
    if (cause instanceof ProtocolProblem) {
      return { ok: false, problem: cause };
    }
    throw cause;
  }
}